src/protocol/check_zero.rs

```diff
@@ -0,0 +1,149 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::fabric::Network,
+    protocol::{context::ProtocolContext, reveal::reveal, RecordId},
+    secret_sharing::Replicated,
+};
+
+use serde::{Deserialize, Serialize};
+
+/// A message sent by each helper when they've multiplied their own shares
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct DValue<F> {
+    d: F,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Step {
+    MultiplyWithR,
+    RevealR,
+}
+
+impl crate::protocol::Step for Step {}
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::MultiplyWithR => "multiply_with_r",
+            Self::RevealR => "reveal_r",
+        }
+    }
+}
+
+/// A very simple protocol to check if a replicated secret sharing is in fact a sharing of zero
+///
+/// NOTE: this protocol leaks information about `v` the helpers. Please only use this in cases where
+/// this type of information leakage is acceptable, such as where `v` is the product of a secret value
+/// and a random, unknown value.
+///
+/// This is an implementation of PROTOCOL 3.7 from the paper:
+/// Fast Large-Scale Honest-Majority MPC for Malicious Adversaries
+/// <https://link.springer.com/content/pdf/10.1007/978-3-319-96878-0_2.pdf>
+///
+/// The parties start out holding a replicated secret sharing of a value `v`, which they would like to check for equality to zero.
+/// First, the parties generate a secret sharing of a random value `r`, which is not known to any of them.
+/// Next, the parties compute a secret sharing of `r * v` using a multiplication protocol that is secure up to an additive attack.
+/// Then, the parties reveal (sometimes called "open") the secret sharing of `r * v`
+/// If `v` was a secret sharing of zero, this revealed value will also be zero.
+/// On the other hand, if `v` was NOT a secret sharing of zero, then there are two possibilities:
+/// 1.) If the randomly chosen value `r` just so happenned to be zero, then the revealed value will be zero.
+/// This will happen with probability `1/|F|` (where `|F|` denotes the cardinality of the field)
+/// 2.) If the randomly chosen value `r` is any other value in the field aside from zero, then the revealed value will NOT be zero.
+///
+/// Clearly, the accuracy of this protocol is highly dependent on the field that is used.
+/// In a large field, like the integers modulo 2^31 - 1, the odds of `check_zero` returning "true"
+/// when `v` is NOT actually a sharing of zero is extrmely small; it is less than one two billion odds.
```

Comment:
```suggestion
/// when `v` is NOT actually a sharing of zero is extremely small; it is less than one two billion odds.
```

This is not quite right.  The probability that this returns true, based on the above given $1-(1-\frac{1}{|F|})^2$.  For $|F|=2^{31}-1$ I think that is one in a billion, which is basically $1:2^{30}$.  That's still pretty good, to be clear, but you are off by a factor of two.

Reply:
I don't think so. Why are you squaring it?

The only way it goes wrong is if r == 0. The odds of choosing that is just one out of the size of the field.

Here's the relevant snippet from the paper:

"`F_checkZero` receives the honest parties shares, and use them to reconstruct the
shared value. Then, if it is 0, then the simulator decides whether to send accept
to the honest parties or reject. Otherwise, `F_checkZero` tosses a coin to decide what
to send to the parties (i.e., in this case, the simulator is not given the opportunity
to modify the output). In particular, when the checked value does not equal to 0,
the output will still be accept with probability 1 / |F| . This captures the event in
Protocol 3.7 where v != 0 but T = 0, which happens also with probability 1/F 
since r is uniformly distributed over F."

Reply:
I'm thinking about both v and r being random, which leads to the squaring.  But I guess that your point would be that - in this specific application of `F_checkZero` - the value of `v` cannot be zero.  I didn't see how that was possible in the case where you are doing a dot product.  Though each individual value is non-zero, the sum is not guaranteed to be.

Reply:
Ah, I think you're conflating two separate things:

This protocol is just "check zero". For this particular functionality, if "v" is indeed zero, then this protocol will always return the correct answer. Put another way, there are no "false negatives".
If "v" is NOT zero, this protocol MIGHT return an incorrect answer. The odds of that are just the odds that r == 0.

When we later on get to the application of this protocol to upgrading a circuit to malicious security, THEN, indeed, the odds are different, because we are talking about something else. Then, the odds that the protocol says "Everything is all good!" when everything is NOT all good are indeed 2/ |F|.

Quoting from the paper: 
"Overall, we therefore have that the adversary can cheat with probability at most 2/|F|. We prove this formally in Lemma 4.2. A full specification appears in Protocol 4.1"

...but this will be in a different comment in a different file. If you're just reading the "check_zero.rs" file, the current comment is accurate. I will definitely have another comment, in whatever that other file is eventually called, explaining the odds of successfully cheating.